ledger/sandbox/src/test/lib/scala/com/digitalasset/platform/store/dao/JdbcLedgerDaoTransactionTreesSpec.scala

```diff
@@ -0,0 +1,158 @@
+          created.observers should contain theSameElementsAs createNode.stakeholders.diff(
+            createNode.signatories)
+          created.templateId shouldNot be(None)
+      }
+    }
+  }
+
+  it should "return the expected flat transaction for a correct request (exercise)" in {
+    for {
+      (_, create) <- storeCreateTransaction()
+      created <- ledgerDao.transactionsReader
+        .lookupTransactionTreeById(create.transactionId, Set(create.submittingParty.get))
+      (offset, exercise) <- storeExerciseTransaction(
+        AbsoluteContractId(created.get.transaction.get.eventsById.head._2.getCreated.contractId))
+      result <- ledgerDao.transactionsReader
+        .lookupTransactionTreeById(exercise.transactionId, Set(exercise.submittingParty.get))
+    } yield {
+      inside(result.value.transaction) {
+        case Some(transaction) =>
+          val (eventId, exerciseNode: NodeExercises.WithTxValue[EventId, AbsoluteContractId]) =
+            exercise.transaction.nodes.head
+          transaction.commandId shouldBe exercise.commandId.get
+          transaction.offset shouldBe ApiOffset.toApiString(offset)
+          transaction.transactionId shouldBe exercise.transactionId
+          transaction.workflowId shouldBe exercise.workflowId.getOrElse("")
+          val exercised = transaction.eventsById.values.loneElement.getExercised
+          transaction.rootEventIds.loneElement shouldEqual exercised.eventId
+          exercised.eventId shouldBe eventId
+          exercised.witnessParties should contain only exercise.submittingParty.get
+          exercised.contractId shouldBe exerciseNode.targetCoid.coid
+          exercised.templateId shouldNot be(None)
+          exercised.actingParties should contain theSameElementsAs exerciseNode.actingParties
+          exercised.childEventIds shouldBe Seq.empty
+          exercised.choice shouldBe exerciseNode.choiceId
+          exercised.choiceArgument shouldNot be(None)
+          exercised.consuming shouldBe true
+          exercised.exerciseResult shouldNot be(None)
+      }
+    }
+  }
+
+  it should "return the expected flat transaction for a correct request (create, exercise)" in {
+    for {
+      (offset, tx) <- storeFullyTransientTransaction()
+      result <- ledgerDao.transactionsReader
+        .lookupTransactionTreeById(tx.transactionId, Set(tx.submittingParty.get))
+    } yield {
+      inside(result.value.transaction) {
+        case Some(transaction) =>
+          val (createEventId, createNode) =
+            tx.transaction.nodes.collectFirst {
+              case (eventId, node: NodeCreate.WithTxValue[AbsoluteContractId]) =>
+                eventId -> node
+            }.get
+          val (exerciseEventId, exerciseNode) =
+            tx.transaction.nodes.collectFirst {
+              case (eventId, node: NodeExercises.WithTxValue[EventId, AbsoluteContractId]) =>
+                eventId -> node
+            }.get
+
```

Comment:
You could additionally check that the order of `root_event_ids` is the same as `[createEventId, exerciseEventId]`.

Reply:
Yes, thanks, copy-pasta.

Reply:
:smile:

Reply:
> You could additionally check that the order of `root_event_ids` is the same as `[createEventId, exerciseEventId]`.

Here it's not really relevant, here I'm extracting the events from the transaction created originally, the order for the response is checked a few lines below.